### ISSUE TITLE
Possible inspect inconsistency for certain container images

### ISSUE THREAD

**Issue by pditommaso** [seqeralabs/wave#604]
There may be an inconsistency in the output of the inspect command for some containers. 

For example 

```
» wave --inspect -i quay.io/biocontainers/rseqc:3.0.1--py37h516909a_1 | jq
{
  "container": {
    "config": {
      "architecture": "amd64",
      "config": {
        "attachStderr": false,
        "attachStdin": false,
        "attachStdout": false,
        "cmd": [
          "/bin/sh"
        ],
        "entrypoint": [],
        "env": [
          "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        ],
        "tty": false
      },
      "container": "0670a689f8e0c8da4a183d000d033cc2ad9b65c40936ba1149d7a17c4d7c1106",
      "created": "2019-10-27T20:02:49.397671191Z",
      "rootfs": {
        "diff_ids": [
          "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef",
          "sha256:461f75075df2a334108df11b185363d2afda21517fda5ffb10852020e7639b7b",
          "sha256:d9c122ba3e7627ce3008fd063788e8a1212d8de86248d66772f39f7b978634ef",
          "sha256:2e4090b968e1df4c6c6d8024c7e60f81148cf4b4dacaa2ece4f3e99f3dba3fc8",
          "sha256:4266fd267d6bfee846b1abef4ab786955e2d1625354fc59ed2ff25a0461db2d7",
          "sha256:b46c2431d956fdcc4b659bbb008e9df0a1ee186088d494267232e562b3b7f3d6",
          "sha256:fe8ed9f638c2465e53d616918ce063f2df83c9ce5cb2a18ee5f334432c1dc76f",
          "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef",
          "sha256:b7a67dc9db8baf3fe750b0161bb0e57544960fdbcfdc61a4e68d8c4c0a43623c",
          "sha256:543c27ce6552e277358445e1d78f3bf725cb6dde5057de080dd2d09d9430e41c"
        ],
        "type": "layers"
      }
    },
    "digest": "sha256:837221f4cb062588b8c7951d59a51f06237563eaef17af965de2fdc68f7e8f18",
    "hostName": "https://quay.io",
    "imageName": "biocontainers/rseqc",
    "manifest": {
      "config": {
        "digest": "sha256:8e8d841718c7e0b349a6d0a1f0a017938cf80eaa27d8a2be489f19bc3c11e42a",
        "mediaType": "application/vnd.docker.container.image.v1+json",
        "size": 3571
      },
      "layers": [
        {
          "digest": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 32,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
        },
        {
          "digest": "sha256:77c6c00e8b61bb628567c060b85690b0b0561bb37d8ad3f3792877bddcfe2500",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 2170406,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:77c6c00e8b61bb628567c060b85690b0b0561bb37d8ad3f3792877bddcfe2500"
        },
        {
          "digest": "sha256:3aaade50789a6510c60e536f5e75fe8b8fc84801620e575cb0435e2654ffd7f6",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 247,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:3aaade50789a6510c60e536f5e75fe8b8fc84801620e575cb0435e2654ffd7f6"
        },
        {
          "digest": "sha256:00cf8b9f3d2a08745635830064530c931d16f549d031013a9b7c6535e7107b88",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 220,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:00cf8b9f3d2a08745635830064530c931d16f549d031013a9b7c6535e7107b88"
        },
        {
          "digest": "sha256:7ff999a2256f84141f17d07d26539acea8a4d9c149fefbbcc9a8b4d15ea32de7",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 2533,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:7ff999a2256f84141f17d07d26539acea8a4d9c149fefbbcc9a8b4d15ea32de7"
        },
        {
          "digest": "sha256:d2ba336f2e4458a9223203bf17cc88d77e3006d9cbf4f0b24a1618d0a5b82053",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 33694,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:d2ba336f2e4458a9223203bf17cc88d77e3006d9cbf4f0b24a1618d0a5b82053"
        },
        {
          "digest": "sha256:dfda3e01f2b637b7b89adb401f2f763d592fcedd2937240e2eb3286fabce55f0",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 243314,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:dfda3e01f2b637b7b89adb401f2f763d592fcedd2937240e2eb3286fabce55f0"
        },
        {
          "digest": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 32,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
        },
        {
          "digest": "sha256:10c3bb32200bdb5006b484c59b5f0c71b4dbab611d33fca816cd44f9f5ce9e3c",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 873691,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:10c3bb32200bdb5006b484c59b5f0c71b4dbab611d33fca816cd44f9f5ce9e3c"
        },
        {
          "digest": "sha256:64278c93a0279818d7161715cf7361caeb666eb8763ac562d02369b4dbf60fed",
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 299826830,
          "uri": "https://quay.io/v2/biocontainers/rseqc/blobs/sha256:64278c93a0279818d7161715cf7361caeb666eb8763ac562d02369b4dbf60fed"
        }
      ],
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "schemaVersion": 2
    },
    "reference": "3.0.1--py37h516909a_1",
    "registry": "quay.io"
  }
}

```

Instead using Skopeo the following is displayed 

```
» skopeo inspect docker://quay.io/biocontainers/rseqc:3.0.1--py37h516909a_1
{
    "Name": "quay.io/biocontainers/rseqc",
    "Digest": "sha256:837221f4cb062588b8c7951d59a51f06237563eaef17af965de2fdc68f7e8f18",
    "RepoTags": [
      ... 
    ],
    "Created": "2019-10-27T20:02:49.397671191Z",
    "DockerVersion": "3.0.6",
    "Labels": {},
    "Architecture": "amd64",
    "Os": "linux",
    "Layers": [
        "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4",
        "sha256:77c6c00e8b61bb628567c060b85690b0b0561bb37d8ad3f3792877bddcfe2500",
        "sha256:3aaade50789a6510c60e536f5e75fe8b8fc84801620e575cb0435e2654ffd7f6",
        "sha256:00cf8b9f3d2a08745635830064530c931d16f549d031013a9b7c6535e7107b88",
        "sha256:7ff999a2256f84141f17d07d26539acea8a4d9c149fefbbcc9a8b4d15ea32de7",
        "sha256:d2ba336f2e4458a9223203bf17cc88d77e3006d9cbf4f0b24a1618d0a5b82053",
        "sha256:dfda3e01f2b637b7b89adb401f2f763d592fcedd2937240e2eb3286fabce55f0",
        "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4",
        "sha256:10c3bb32200bdb5006b484c59b5f0c71b4dbab611d33fca816cd44f9f5ce9e3c",
        "sha256:64278c93a0279818d7161715cf7361caeb666eb8763ac562d02369b4dbf60fed"
    ],
    "LayersData": [
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4",
            "Size": 32,
            "Annotations": null
        },
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:77c6c00e8b61bb628567c060b85690b0b0561bb37d8ad3f3792877bddcfe2500",
            "Size": 2170406,
            "Annotations": null
        },
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:3aaade50789a6510c60e536f5e75fe8b8fc84801620e575cb0435e2654ffd7f6",
            "Size": 247,
            "Annotations": null
        },
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:00cf8b9f3d2a08745635830064530c931d16f549d031013a9b7c6535e7107b88",
            "Size": 220,
            "Annotations": null
        },
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:7ff999a2256f84141f17d07d26539acea8a4d9c149fefbbcc9a8b4d15ea32de7",
            "Size": 2533,
            "Annotations": null
        },
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:d2ba336f2e4458a9223203bf17cc88d77e3006d9cbf4f0b24a1618d0a5b82053",
            "Size": 33694,
            "Annotations": null
        },
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:dfda3e01f2b637b7b89adb401f2f763d592fcedd2937240e2eb3286fabce55f0",
            "Size": 243314,
            "Annotations": null
        },
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4",
            "Size": 32,
            "Annotations": null
        },
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:10c3bb32200bdb5006b484c59b5f0c71b4dbab611d33fca816cd44f9f5ce9e3c",
            "Size": 873691,
            "Annotations": null
        },
        {
            "MIMEType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "Digest": "sha256:64278c93a0279818d7161715cf7361caeb666eb8763ac562d02369b4dbf60fed",
            "Size": 299826830,
            "Annotations": null
        }
    ],
    "Env": [
        "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
    ]
}
```

Weird facts: 
1. the `diff_ids` reported by Wave does not match the layers checksum 
2. in layers, both for Wave and Skopeo, there are multiple occurrence of the same checksum e.g. `a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4`

**Comment by pditommaso** [comment]
@fntlnz what's your take on this ☝️ . Independently of the tool used (ie. wave vs skopeo) what I found very weird that there multiple layers with the same checksum. Is that valid?

**Comment by fntlnz** [comment]
Regarding the duplicated layers those SHA256 are related to empty layers

`5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef` is the "legacy" empty layer sha, https://github.com/moby/moby/blob/2c0100fbde612ecbaae3a08f960ae726aa57ecc7/layer/empty.go#L12


While `a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4` is the sha of the "new" empty layer blob, see https://github.com/distribution/distribution/blob/749f6afb4572201e3c37325d0ffedb6f32be8950/manifest/schema1/config_builder.go#L22-L29

So in short, wave is using the "legacy" blob, skopeo is using the "new" blob that's why you see different shasums for the same thing.

**Comment by fntlnz** [comment]
Edit: deleted my previous reply because I wanted to dig deeper and it wasn't very precise.

From my understanding

`5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef` is the empty layer sha, https://github.com/moby/moby/blob/2c0100fbde612ecbaae3a08f960ae726aa57ecc7/layer/empty.go#L12

We can verify this by doing

```bash
dd if=/dev/zero of=empty_file bs=1024 count=1
tar -cf empty.tar empty_file
sha256sum empty_file                                                                                                                    
5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef  empty_file
```


While `a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4` is the same thing but gzipped so it has a different sha https://github.com/distribution/distribution/blob/749f6afb4572201e3c37325d0ffedb6f32be8950/manifest/schema1/config_builder.go#L22-L29

This can be verified by downloading one of those layers



```
wget https://quay.io/v2/biocontainers/rseqc/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 

```

And then checking if it has `5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef` as its content sha256
```
mv sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 file.tar.gz
gunzip file.tar.gz
sha256sum file.tar                                                                                                                              
5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef  file.tar
```
Yes it does :partying_face: 

Here is the raw manifest list returned by quay for that image via curl which is the same as what we can find in both skopeo and wave.

```bash
curl https://quay.io/v2/biocontainers/rseqc/manifests/sha256:837221f4cb062588b8c7951d59a51f06237563eaef17af965de2fdc68f7e8f18 GET
```

<details>
<summary>expand to see manifest</summary>

```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
   "config": {
      "mediaType": "application/vnd.docker.container.image.v1+json",
      "size": 3571,
      "digest": "sha256:8e8d841718c7e0b349a6d0a1f0a017938cf80eaa27d8a2be489f19bc3c11e42a"
   },
   "layers": [
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 32,
         "digest": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 2170406,
         "digest": "sha256:77c6c00e8b61bb628567c060b85690b0b0561bb37d8ad3f3792877bddcfe2500"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 247,
         "digest": "sha256:3aaade50789a6510c60e536f5e75fe8b8fc84801620e575cb0435e2654ffd7f6"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 220,
         "digest": "sha256:00cf8b9f3d2a08745635830064530c931d16f549d031013a9b7c6535e7107b88"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 2533,
         "digest": "sha256:7ff999a2256f84141f17d07d26539acea8a4d9c149fefbbcc9a8b4d15ea32de7"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 33694,
         "digest": "sha256:d2ba336f2e4458a9223203bf17cc88d77e3006d9cbf4f0b24a1618d0a5b82053"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 243314,
         "digest": "sha256:dfda3e01f2b637b7b89adb401f2f763d592fcedd2937240e2eb3286fabce55f0"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 32,
         "digest": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 873691,
         "digest": "sha256:10c3bb32200bdb5006b484c59b5f0c71b4dbab611d33fca816cd44f9f5ce9e3c"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 299826830,
         "digest": "sha256:64278c93a0279818d7161715cf7361caeb666eb8763ac562d02369b4dbf60fed"
      }
   ]
}
```

</details>


So the list of layers is correct in all of them, and both wave and skopeo are reporting the layers correctly.

The thing you are seeing here is that wave is reporting both the gzipped shasums in `layers` and the non gzipped ones in `diff_ids` but based on what I see I would say that the wave output is consistent if seeing the non gzipped layer shas is the desired behavior.


We can do the same exercise as above with any other blob, let's say `77c6c00e8b61bb628567c060b85690b0b0561bb37d8ad3f3792877bddcfe2500` which is at position 2 and position 2 corresponds to `461f75075df2a334108df11b185363d2afda21517fda5ffb10852020e7639b7b` in `diff_ids`


```bash
wget https://quay.io/v2/biocontainers/rseqc/blobs/sha256:77c6c00e8b61bb628567c060b85690b0b0561bb37d8ad3f3792877bddcfe2500
mv sha256:77c6c00e8b61bb628567c060b85690b0b0561bb37d8ad3f3792877bddcfe2500 file.tar.gz

Now, if we check the shasum we get exactly `461f75075df2a334108df11b185363d2afda21517fda5ffb10852020e7639b7b` as expected!
```bash
sha256sum file.tar                                                                                                                        
461f75075df2a334108df11b185363d2afda21517fda5ffb10852020e7639b7b  file.tar
```

**Comment by pditommaso** [comment]
Thanks for this detailed reporting. Wave retrieves the `diff_ids` along with the image config [here](https://github.com/seqeralabs/wave/blob/c90411f0de4ba2716d3305a1c7ead1e6fc2bba19/src/main/groovy/io/seqera/wave/core/ContainerAugmenter.groovy#L535-L539). 

For some reason I was expecting the `diff_ids` to be the same as the layers checksum. 

We need to conclude that's fine.
